MP Connect dialog overhaul and server management UI for #4564

shikadiqueen opened the pull request:
There has been functionality to manage custom servers in the player's preferences file since as far back as commit 5fb0b2a91123474d147c9de4918fc11a1b4c15ac in June 2007, but there's never been a user interface for it.

![image](https://user-images.githubusercontent.com/489895/68990311-a7822c80-0830-11ea-85d1-ce1dcec9db16.png)

This commit removes the current server list dialog attached to the MP Connect dialog and replaces it with the same functionality embedded into the latter, as well as the ability to edit the server list in the player's preferences file. There's some code movement in a few other places since the previous function in the preferences namespace doesn't suit the new code (it always returned a combined list of game and player-defined servers).

There's some disabled code referencing a situation where the listbox with the list of servers allows the code (and the player) to clear the selection. It's disabled because there are some UX issues with it noted in the code comments associated to it.

Fixes #4564.

Vultraz commented on src/gui/dialogs/edit_text.cpp:
```diff
@@ -41,14 +46,37 @@ namespace dialogs
+   button& ok_button = find_widget<button>(get_window(), "ok", false);
+
+   auto clean_value = text.get_value();
+	boost::trim(clean_value);
```
You don't need to create a temporary here. You can use `boost::trim_copy`.

AI0867 replied:
The temporary is still created, though it is no longer used.

Vultraz commented on src/gui/dialogs/edit_text.cpp:
```diff
@@ -41,14 +46,37 @@ namespace dialogs
+
+void edit_text::pre_show(window& window)
+{
+   if(disallow_empty_) {
```
Bunch of spaces indent around here.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
+}
+#endif
+
+std::string clean_textbox_value(text_box& text_box)
```
You don't need this, you can use `boost::trim_copy`.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
+void mp_connect::select_first_match()
+{
+	window* window = get_window();
+	if(!window) {
```
You don't need to always check `window` is valid. It's guaranteed to exist as long as the dialog is shown.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
+void mp_connect::on_server_add()
+{
+	window* window = get_window();
+	if(!window) {
```
Ditto about checking `window`.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
 {
-	if(get_retval() == retval::OK) {
+	window* window = get_window();
+	if(!window) {
```
Ditto about checking `window`.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
 {
-	assert(host_name);
+	window* window = get_window();
+	if(!window) {
```
Ditto about checking `window`.

Vultraz commented on src/preferences/game.cpp:
```diff
@@ -380,28 +380,43 @@ void _set_lobby_joins(int show)
+		game_config::server_info sinf;
+		sinf.name = server["name"].str();
+		sinf.address = server["address"].str();
+		pref_servers.push_back(sinf);
```
You can use `emplace_back` here.

shikadiqueen replied:
There's a (redundant) default constructor explicitly defined for `game_config::server_info` which interferes with this and other things such as move assignment.

Vultraz commented on src/gui/dialogs/multiplayer/mp_connect.cpp:
```diff
@@ -16,174 +16,304 @@
 {
-	set_restore(true);
+	window* window = get_window();
+	if(!window) {
```
Ditto about checking window.

shikadiqueen commented:
Addressed all issues except for the `emplace_back` recommendation since the class' custom default ctor gets in the way. I may append that to the series later today, though.

shikadiqueen commented:
Sorry, that wasn't quite right. I forgot a `boost::trim` in the edit text dialog.

shikadiqueen commented:
My mistake, the `emplace_back` issue is simply because the class doesn't define a constructor that takes two strings as arguments. Forcing use of the empty ctor and working on the result is a viable alternative though.

CelticMinstrel commented:
Looks good, would be nice if we can solve the problems with allowing an empty selection in the list though. If I understand correctly, what happened was you type in a server that's not in the list, which then causes the selection to be cleared, which then means … what exactly? Clicking the plus button with an empty selection should probably append to the end, but it sounds like you want it to insert where the selection just was? In that case, maybe you could just keep the current selection if the newly-typed server doesn't match.

shikadiqueen commented:
> If I understand correctly, what happened was you type in a server that's not in the list, which then causes the selection to be cleared, which then means … what exactly? Clicking the plus button with an empty selection should probably append to the end, but it sounds like you want it to insert where the selection just was? In that case, maybe you could just keep the current selection if the newly-typed server doesn't match.

Personally I believe that it makes sense to always show where the selection is at when it comes to these kind of things. If the selection is cleared then you're assuming that the player will remember where it was at when you perform an operation next while you keep track of it behind the scenes. For me, any situation where the user interface *knows* things that aren't being visibly displayed is an attack on the user.

Now, technically, adding right below the current/last selection isn't a strictly necessary feature. Appending always to the end of the list would work just as well, especially since this isn't going to see as much use as for people to wind up in a situation where having to rearrange the list by deleting and re-adding entries is going to become a concrete problem. We could even have up/down buttons to rearrange list items, stealing more real estate and making the UI seem cramped.

I'd like to think that the decision I made (there's always a selection even if it doesn't match) is a better compromise than the alternative from a UX point of view, since ultimately the authoritative source of the server address is always going to be whatever is displayed on the textbox.

CelticMinstrel commented:
The only situation where I think I might disagree is when you first open the window and the remembered server doesn't match anything on the list.

shikadiqueen commented:
That doesn't seem important to me.

On the other hand, if you feel inclined to write code to disallow the player form clearing the selection themselves by clicking on it that would be a reasonable compromise.

CelticMinstrel commented:
Yeah… well, I'll think about it. If I feel like it, maybe I'll do that at some point after this is merged.
